src/dynamic_programming/python/coins/src/coin_change.py

```diff
@@ -1,28 +1,87 @@
 """
-Module: coin_change
+This module provides functions to find the minimum number of coins needed to make the given amount of money using the available coins.
+"""
+
+__author__ = "Abed Hamami"
 
-This module provides a function for calculating the minimum number of coins needed to make up a given amount.
 
-Author: Abed Hamami
-"""
 def coin_change_basic(amount, coins):
-   """
-   Calculates the minimum number of coins needed to make up a given amount.
-
-   Parameters:
-       amount (int): The target amount to make up.
-       coins (list): The list of available coin denominations.
-
-   Returns:
-       int: The minimum number of coins needed to make up the amount. Returns -1 if it is not possible to make up the amount using the given coins.
-   """
-   if amount == 0:
-       return 0
-   if amount < 0:
-       return -1
-   res = float('inf')
-   for i in range(len(coins)):
-       sub_res = coin_change_basic(amount - coins[i], coins)
-       if sub_res >= 0 and sub_res < res:
-           res = sub_res + 1
-   return -1 if res == float('inf') else res
\ No newline at end of file
+    """
+    Calculates the minimum number of coins needed to make a given amount of change.
+
+    Args:
+        amount (int): The amount of change to be made.
+        coins (List[int]): A list of coin denominations available.
+
+    Returns:
+        int: The minimum number of coins needed to make the given amount of change. Returns -1 if it is not possible to make the change.
+    """
+    if amount == 0:
+        return 0
+    if amount < 0:
+        return -1
+    res = float("inf")
+    for coin in coins:
+        sub_res = coin_change_basic(amount - coin, coins)
+        if sub_res >= 0 and sub_res < res:
+            res = sub_res + 1
+    return -1 if res == float("inf") else res
+
+
+def coin_change_memo(amount, coins):
+    """
+    Calculates the minimum number of coins needed to make up a given amount using a dynamic programming approach with memoization.
+
+    Parameters:
+        amount (int): The target amount to be made up using the coins.
+        coins (list): A list of coin denominations.
+
+    Returns:
+        int: The minimum number of coins needed to make up the target amount. Returns -1 if it is not possible to make up the target amount using the given coins.
+    """
+    memo = {}
+
+    def dp(n):
```

Comment:
Please provide a more descriptive function name(make sure it's in snake_case)

Reply:
yea good catch! I'll update that one!